src/ucs/datastruct/ptr_array.c

```diff
@@ -54,6 +54,25 @@ ucs_ptr_array_freelist_set_next(ucs_ptr_array_elem_t *elem, unsigned next)
                     (((ucs_ptr_array_elem_t)next) << UCS_PTR_ARRAY_NEXT_SHIFT);
 }
 
+static inline unsigned
+ucs_ptr_array_freelist_get_next_params(ucs_ptr_array_elem_t elem, uint32_t *size_free)
+{
+    ucs_assert(__ucs_ptr_array_is_free(elem));
+    *size_free = elem >> UCS_PTR_ARRAY_SIZE_FREE_SHIFT;
+    return ((elem & UCS_PTR_ARRAY_NEXT_MASK) >> UCS_PTR_ARRAY_NEXT_SHIFT);
+}
+
+static inline void
+ucs_ptr_array_freelist_element_set_free(ucs_ptr_array_elem_t *elem, uint32_t size_free,
+	unsigned next)
+{
+    ucs_assert(next <= UCS_PTR_ARRAY_NEXT_MASK);
+
```

Comment:
In the case of grow, we iterate through the list from the first to the last but the i+1 index is not free it (will be free in the next iteration).
It might make sense to put this assert in the remove() function:

void ucs_ptr_array_remove(ucs_ptr_array_t *ptr_array, unsigned index,
                          uint32_t placeholder)
{
...
    /* Make sure the next element is free */
    ucs_assert(__ucs_ptr_array_is_free((ptr_array)->start[index+size_free_ahead-1]));


Fixed.